tao_compiler/mlir/disc/transforms/disc_transform_schedule.h

```diff
@@ -52,59 +55,136 @@ class PatternDescription {
+};
+
+using ScheduleFactoryPtr = std::unique_ptr<ScheduleFactory>;
 
 // A registry for different schedule factories.
 class ScheduleFactoryRegistry {
  public:
   // Returns the singleton
   static ScheduleFactoryRegistry& get();
 
-  // Inserts the new `ScheduleFactory`. Returns true if inserted, otherwise
-  // false.
-  bool registerScheduleFactory(PatternKind kind, const std::string& tag,
-                               ScheduleFactory);
+  // Returns next available id for schedule factory.
+  static int64_t getNextUniqueId();
 
-  // Returns the schedule factory factor according to `kind` and `tag`.
+  // Inserts the new `ScheduleFactory`. Returns true if inserted, otherwise
+  // false. The larger the `prioirty`, the larger the change the factory being
+  // choosed. Note that we do not allow to assign the same priority for two
+  // different factories with the same pattern kind.
+  bool registerScheduleFactory(PatternKind kind, int priority,
+                               ScheduleFactoryPtr factory);
+  void unregisterScheduleFactory(PatternKind kind, int priority);
+
+  // Returns the schedule factory with the highes priority for `pd`.
   // Returns nullptr if not found.
-  ScheduleFactory getScheduleFactory(PatternKind kind, const std::string& tag);
+  ScheduleFactory* getScheduleFactoryWithHighestPriority(
+      PatternDescription& pd);
+
+  // Returns all suitable schedule factories for `pd`. The returned factory list
+  // is sorted by `priority`. The first one has the highest priority.
+  SmallVector<ScheduleFactory*> getAllCandidateScheduleFactories(
+      PatternDescription& pd);
 
  private:
   ScheduleFactoryRegistry() = default;
-  std::unordered_map<PatternKind,
-                     std::unordered_map<std::string, ScheduleFactory>>
-      factoryMap_;
+  std::unordered_map<PatternKind, std::map<int, ScheduleFactoryPtr>>
+      patternMap_;
 };
 
 // Macros used to define disc transform schedule factory.
-#define DISC_TRANSFORM_SCHEDULE(kind, tag, ...) \
-  DISC_TRANSFORM_SCHEDULE_UNIQ_HELPER(kind, tag, __COUNTER__, __VA_ARGS__)
-
-#define DISC_TRANSFORM_SCHEDULE_UNIQ_HELPER(kind, tag, ctr, ...) \
-  DISC_TRANSFORM_SCHEDULE_UNIQ(kind, tag, ctr, __VA_ARGS__)
-
-#define DISC_TRANSFORM_SCHEDULE_UNIQ(kind, tag, ctr, ...) \
-  static bool unused_ret_val_##ctr =                      \
-      ::mlir::disc_ral::ScheduleFactoryRegistry::get()    \
-          .registerScheduleFactory(kind, tag, __VA_ARGS__);
+#define DISC_TRANSFORM_SCHEDULE(kind, priority, T, ...)               \
+  DISC_TRANSFORM_SCHEDULE_UNIQ_HELPER(__COUNTER__, kind, priority, T, \
+                                      __VA_ARGS__)
+
+#define DISC_TRANSFORM_SCHEDULE_UNIQ_HELPER(ctr, kind, priority, T, ...) \
+  DISC_TRANSFORM_SCHEDULE_UNIQ(ctr, kind, priority, T, __VA_ARGS__)
+
+#define DISC_TRANSFORM_SCHEDULE_UNIQ(ctr, kind, priority, T, ...)             \
+  static bool unused_ret_val_##ctr =                                          \
```

Comment:
nit: need to check if register successs?

Reply:
done